File: PracticeAssignment2Part1/Q11_IsPalindrome.java
package PracticeAssignment2Part1;

public class Q11_IsPalindrome {
    /**
     * 
     * assuming tail node is not provided!
     */
    class Node{
        int data;
        Node next;
        Node prev;
        Node(int data){
            this.data = data;
            next =null;
            prev = null;
        }
    }
    /**
     * i will have two pointer one at the front and one at the end. Since its a doubly linked list i can go both ways.
     * i will check if front does not equal end if that conidtion is true i return false.
     * for odd length i can check if both pointers are equal and exit the loop
     * for even length i have added another check checking if the next pointer for front in the end node. this prevents skip when the lsit size is even;
     * time o(n) space o(1)
     * 
     */
    public boolean isPalindrome(Node root){
        if(root==null){
            return true;
        }
        Node tailNode = root;
        while(tailNode.next!=null){
            tailNode = tailNode.next;
        }
        Node frontNode = root;
        while(frontNode != tailNode &&frontNode.next != tailNode){ //second check is so for even size it doesnt skip both pointer;
            if(frontNode.data != tailNode.data){
                return false;
            }
            frontNode = frontNode.next;
            tailNode = tailNode.prev;
        }
        return true;
    }
}

File: delight_oti/q11_IsPalindrome.py
# Technique: Doubly linked list forward-backward two-pointer
# Time Complexity: O(n)
# Space Complexity: O(1)

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None
    
    def isPalindrome(self):

        if not self:
            return True
        
        right = self
        while right.next:
            right = right.next
        
        left = self

        while left != right and left.prev != right:
            if left.data != right.data:
                return False
            
            left = left.next
            right = right.prev

        return True
    
# head = Node(1)
# n2 = Node(2)
# n3 = Node(3)

# head.next = n2
# n2.prev = head

# n2.next = n3
# n3.prev = n2

# print(head.isPalindrome())  # False